test/addons-abi/test_instanceof/test0.js

```diff
@@ -0,0 +1,118 @@
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+// Adapted from the V8 test suite for the instanceof operator to test
+// whether it is possible to implement instanceof using NAPI.
+
+require('../../common');
+var assert = require('assert');
+var addon = require('./build/Release/test_instanceof');
+
+function assertTrue(assertion) {
+	return assert.strictEqual(true, assertion);
+}
+
+function assertFalse(assertion) {
+	assert.strictEqual(false, assertion);
+}
+
+function assertEquals(leftHandSide, rightHandSide) {
+	assert.equal(leftHandSide, rightHandSide);
+}
+
+function assertThrows(statement) {
+	assert.throws(function() {
+		eval(statement);
+	}, Error);
+}
+
+assertTrue(addon.doInstanceOf({}, Object));
+assertTrue(addon.doInstanceOf([], Object));
+
+assertFalse(addon.doInstanceOf({}, Array));
+assertTrue(addon.doInstanceOf([], Array));
+
+function TestChains() {
+  var A = {};
+  var B = {};
+  var C = {};
+  B.__proto__ = A;
+  C.__proto__ = B;
+
+  function F() { }
+  F.prototype = A;
+  assertTrue(addon.doInstanceOf(C, F));
+  assertTrue(addon.doInstanceOf(B, F));
+  assertFalse(addon.doInstanceOf(A, F));
+
+  F.prototype = B;
+  assertTrue(addon.doInstanceOf(C, F));
+  assertFalse(addon.doInstanceOf(B, F));
+  assertFalse(addon.doInstanceOf(A, F));
+
+  F.prototype = C;
+  assertFalse(addon.doInstanceOf(C, F));
+  assertFalse(addon.doInstanceOf(B, F));
+  assertFalse(addon.doInstanceOf(A, F));
+}
+
+TestChains();
+
+
+function TestExceptions() {
+  function F() { }
+  var items = [ 1, new Number(42),
+                true,
+                'string', new String('hest'),
+                {}, [],
+                F, new F(),
+                Object, String ];
+
+  var exceptions = 0;
+  var instanceofs = 0;
+
+  for (var i = 0; i < items.length; i++) {
+    for (var j = 0; j < items.length; j++) {
+      try {
+        if (addon.doInstanceOf(items[i], items[j])) instanceofs++;
+      } catch (e) {
+        assertTrue(e instanceof TypeError);
+        exceptions++;
+      }
+    }
+  }
+  assertEquals(10, instanceofs);
+  assertEquals(88, exceptions);
+
+  // Make sure to throw an exception if the function prototype
+  // isn't a proper JavaScript object.
+  function G() { }
+  G.prototype = undefined;
+  assertThrows("(addon.doInstanceOf({}, G))");
+}
+
+TestExceptions();
```

Comment:
This seems redundant with the checks in test1.js that are also verifying exceptional cases.